airflow/models/xcom.py

```diff
@@ -497,7 +497,7 @@ def get_many(
         elif dag_ids is not None:
             query = query.filter(cls.dag_id == dag_ids)
 
-        if isinstance(map_indexes, range) and abs(map_indexes.step) == 1:
+        if isinstance(map_indexes, range) and map_indexes.step == 1:
             query = query.filter(cls.map_index >= map_indexes.start, cls.map_index < map_indexes.stop)
```

Comment:
The previous logic was wrong; this optimisation only holds when `step=1`. We could make `step=-1` work (by flipping `start` and `end` conditionally but I decided it’s not worth it since that case is not used anywhere.